src/main/scala/org/broadinstitute/hail/expr/Type.scala

```diff
@@ -328,6 +327,11 @@ case class TDict(elementType: Type) extends Type {
 
   override def genValue: Gen[Annotation] =
     Gen.buildableOf2[Map, String, Annotation](Gen.zip(arbitrary[String], elementType.genValue))
+
+  override def valuesSimilar(a1: Annotation, a2: Annotation, tolerance: Double): Boolean =
+    a1 == a2 || (a1 != null && a2 != null
+      && a1.asInstanceOf[Map[String, _]].zip(a2.asInstanceOf[Map[String, _]])
+      .forall { case ((k1, v1), (k2, v2)) => k1 == k2 && elementType.valuesSimilar(v1, v2, tolerance) })
```

Comment:
AFAIK, we actually need a `join` here, right?

This will return `true` when it should not:
```scala
TDict.valuesSimilar(Map("a" -> 1, "b" -> 2), Map("a" -> 1, "b" -> 2, "c" -> 3), 0.0)
```

It's hard to find a good test case because `a1 == a2` will catch a lot of stuff that's exactly equal. Note that this:

```scala
TDict.valuesSimilar(Map("a" -> 1, "b" -> 2), Map("b" -> 2, "a" -> 1, "c" -> 3), 0.0)
```

will correctly return `false`, but only because the ordering has changed!

We need to do a join on the keys. I wrote something for [checking equality of maps in test cases](https://github.com/hail-is/hail/blob/master/src/main/scala/org/broadinstitute/hail/utils/package.scala#L286-L325). We can just steal that and throw out the error messages. We can also fail fast instead of collecting all the failures.

Also, I think we should skip the `a1 == a2` since right now `valusesSimilar` traverses the structure of the annotation twice.